### PR TITLE
[trainer] fix: normalize sft loss by num_tokens in global batch

### PR DESCRIPTION
**Pull request by wuxibin89**
### What does this PR do?

Normalize sft loss by num_tokens in global batch.

TODO
- [ ] Normalize `pg_loss` and `value_loss` in rl trainer pipeline.

Credit to @techkang in https://github.com/volcengine/verl/pull/3729.

### CODE REVIEW

File: verl/workers/roles/utils/losses.py [hunk 1]
```diff
@@ -35,16 +37,14 @@ def sft_loss(config: ActorConfig, model_output, data: TensorDict, dp_group=None)
         loss_mask = data["loss_mask"]
 
         log_prob_flatten = log_prob.values()
-        cu_seqlens = log_prob.offsets()
         loss_mask_flatten = loss_mask.values()
 
         # left-shift the loss mask by one token to align with log_prob
         loss_mask_flatten = torch.roll(loss_mask_flatten, shifts=-1, dims=0)
```
[the remark this answers is not in the document]
Comment (wuxibin89) [reply verl/workers/roles/utils/losses.py:43]: Last token in each sequence is masked by first loss_mask in next sequence after left shift.
```text             
loss_mask:         [0, 0, 1, 1, 1], [0, 0, 0, 1, 1], [0, 1, 1, 1]
loss_mask_flatten: [0, 1, 1, 1, 0], [0, 0, 1, 1, 0], [1, 1, 1, 0]
```
And the num of valid tokens is not changed after left shift.

File: verl/workers/engine/megatron/transformer_impl.py [hunk 2]
```diff
@@ -381,6 +381,15 @@ def load_checkpoint(
 
     def forward_backward_batch(self, data: TensorDict, loss_function: Callable, forward_only=False) -> Any:
         tu.assign_non_tensor(data, sp_size=self.engine_config.context_parallel_size)
+
+        # compute num_tokens in global batch for loss normalization
+        batch_num_tokens = data["loss_mask"].sum().cuda()
+        torch.distributed.all_reduce(
+            batch_num_tokens, op=torch.distributed.ReduceOp.SUM, group=self.get_data_parallel_group()
+        )
```
[the remark this answers is not in the document]
Comment (wuxibin89) [reply verl/workers/engine/megatron/transformer_impl.py]: batch_num_tokens remain the same after loss_mask left shift.

File: verl/workers/engine/fsdp/transformer_impl.py [hunk 3]
```diff
@@ -476,6 +476,14 @@ def forward_backward_batch(self, data: TensorDict, loss_function: Callable, forw
         # note that the global_batch_size should include data on all the dp
         tu.assign_non_tensor(data, sp_size=self.ulysses_sequence_parallel_size)
 
+        # compute num_tokens in global batch for loss normalization
+        batch_num_tokens = data["loss_mask"].sum().cuda()
+        torch.distributed.all_reduce(
+            batch_num_tokens, op=torch.distributed.ReduceOp.SUM, group=self.get_data_parallel_group()
+        )
```
[the remark this answers is not in the document]
Comment (wuxibin89) [reply verl/workers/engine/fsdp/transformer_impl.py]: batch_num_tokens remain the same after loss_mask left shift.

File: verl/workers/engine/megatron/transformer_impl.py [hunk 4]
```diff
@@ -624,18 +633,13 @@ def postprocess_micro_batch_func(self, output, data: TensorDict, forward_only: b
...
         if loss_function is not None:
             loss, metrics = loss_function(model_output=model_output, data=data, dp_group=self.get_data_parallel_group())
             # scale loss by num_micro_batch because megatron will scale loss
             # by n_micro_batch and cp size inside pp schedule
-            n_micro_batch = data["num_micro_batch"]
-            loss = loss * n_micro_batch / mpu.get_context_parallel_world_size()
-            global_bsz = data["global_batch_size"]
-            loss_scale_factor = local_micro_bsz / (global_bsz / self.get_data_parallel_size())
-            loss = loss * loss_scale_factor
+            loss = loss * data["num_micro_batch"] / mpu.get_context_parallel_world_size()
```
Comment (techkang) [inline verl/workers/engine/megatron/transformer_impl.py:642]: We should add a note: the sft_loss is used in the FSDP backend (which handles scaling automatically), while it requires manual scaling in the Megatron backend.

Comment (wuxibin89) [reply verl/workers/engine/megatron/transformer_impl.py:642]: Add note in `verl/workers/roles/utils/losses.py`
